File: leetcode/dynamicProgramming/maxProductSubArray.go
/*
curr = 4;
max = 6;

*/

func maxProduct(nums []int) int {
	left := 1
	right := 1
	var max int = nums[0]
	for i := 0; i < len(nums); i++ {
		if left == 0 {
			left = 1
		}
		if right == 0 {
			right = 1
		}
		left = left * nums[i]
		right = right * nums[len(nums)-i-1]
		max = Max(max, Max(left, right))
	}
	return max
}
func Max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

File: leetcode/dynamicProgramming/maxProductSubarray.js
/*

2,3,-2,3,-1,2,3,-3
2,6,12,., .,.,., .

[  |     |    |   ]

*/
/**
 * @param {number[]} nums
 * @return {number}
 */
 var maxProduct = function(nums) {
    let max = nums[0];
    let left = 1;
    let right = 1;
    const len = nums.length;
    for(let i = 0; i < len; i++) {
        left *= nums[i];
        right *= nums[len - 1 - i];
        max = Math.max(max, left, right);
        left = left === 0 ? 1: left;
        right = right === 0 ? 1: right;
    }
    return max;
};
